djtriggers/models.py

```diff
@@ -82,28 +90,95 @@ def set_source(self, *args):
-        if logger:
-            self.logger = get_logger(logger)
-        now = timezone.now()
-        if not force and self.date_processed is not None:
-            raise AlreadyProcessedError()
-        if not force and self.process_after and self.process_after >= now:
-            raise ProcessLaterError(self.process_after)
+    @shared_task(filter=task_method)
+    def process(self, force=False, logger=None, dictionary={}, use_statsd=False):
+        """
+        Executes the Trigger
+        :param bool force: force the execution
+        :param string logger: slug of preferred logger
+        :param dict dictionary: dictionary needed by trigger to execute
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        # The task gets locked because multiple tasks in the queue can process the same trigger.
+        # The lock assures no two tasks can process a trigger simultaneously.
+        # The check for date_processed assures a trigger is not executed multiple times.
+        try:
+            trigger_lock_name = 'execute_trigger%d' % self.pk
+            file_lock = FileLock(trigger_lock_name)
+            file_lock.acquire(0)
+        except AlreadyLocked:
+            return
 
         try:
-            self.logger.log_result(self, self._process(dictionary))
-        except ProcessLaterError as e:
-            self.process_after = e.process_after
-            self.save()
-            raise
-        if self.date_processed is None:
-            self.date_processed = now
-        self.save()
+            if logger:
+                self.logger = get_logger(logger)
+            now = timezone.now()
+            if not force and self.date_processed is not None:
+                raise AlreadyProcessedError()
+            if not force and self.process_after and self.process_after >= now:
+                raise ProcessLaterError(self.process_after)
+
+            try:
+                # execute trigger
+                self.logger.log_result(self, self._process(dictionary))
+                self._handle_execution_success(use_statsd)
+            except ProcessLaterError as e:
+                self.process_after = e.process_after
+                self.save()
+                raise
+            except Exception as e:
+                self._handle_execution_failure(e, use_statsd)
+                raise
+        finally:
+            file_lock.release()
+
 
     def _process(self, dictionary):
         raise NotImplementedError()
 
+    def _handle_execution_failure(self, exception, use_statsd=False):
+        """
+        Handle execution failure of the trigger
+        :param Exception exception: the exception raised during failure
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        self.nb_of_tries += 1
+        if self.nb_of_tries >= getattr(settings, 'DJTRIGGERS_TRIGGER_RETRIES_WARNING_LVL', sys.maxint):
```

Comment:
Here is a good read from @kristofclaes on readability: https://kristofclaes.github.io/2016/07/15/code-should-communicate-intent/ It talks about avoiding abbreviations and reducing line count.